packages/react/menu/src/Menu.tsx

```diff
@@ -94,65 +237,179 @@ const MenuContent = React.forwardRef((props, forwardedRef) => {
-     * @defaultValue ltr
-     */
-    dir?: RovingFocusGroupOwnProps['dir'];
+/* ---------------------------------------------------------------------------------------------- */
 
-    /**
-     * Whether keyboard navigation should loop around
-     * @defaultValue false
-     */
-    loop?: RovingFocusGroupOwnProps['loop'];
+type MenuSubContentOwnProps = Omit<
+  Polymorphic.OwnProps<typeof MenuContentImpl>,
+  | keyof MenuContentImplPrivateProps
+  | 'side'
+  | 'portalled'
+  | 'disabledOutsidePointerEvents'
+  | 'disableOutsideScroll'
+  | 'trapFocus'
+  | 'onOpenAutoFocus'
+  | 'onCloseAutoFocus'
+>;
 
-    /**
-     * Whether the `MenuContent` should render in a `Portal`
-     * (default: `true`)
-     */
-    portalled?: boolean;
-  }
+type MenuSubContentPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof MenuContentImpl>,
+  MenuSubContentOwnProps
+>;
+
+const MenuSubContent = React.forwardRef((props, forwardedRef) => {
+  const context = useMenuContext(CONTENT_NAME);
+  const ref = React.useRef<MenuContentElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, ref);
+  return context.isSubmenu ? (
+    <MenuContentImpl
+      align="start"
+      id={context.contentId}
+      {...props}
+      ref={composedRefs}
+      side={context.dir === 'rtl' ? 'left' : 'right'}
+      portalled
+      disableOutsidePointerEvents={false}
+      disableOutsideScroll={false}
+      trapFocus={false}
+      onEntryFocus={context.onEntryFocus}
+      onOpenAutoFocus={(event) => event.preventDefault()}
+      // The menu might close because of focusing another menu item in the parent menu. We
+      // don't want it to refocus the trigger in that case so we handle trigger focus ourselves.
+      onCloseAutoFocus={(event) => event.preventDefault()}
+      onFocusOutside={composeEventHandlers(props.onFocusOutside, (event) => {
+        // We prevent closing when the trigger is focused to avoid triggering a re-open animation
+        // on pointer interaction.
+        if (event.target !== context.trigger) context.onOpenChange(false);
+      })}
+      onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
+        // We need to explicitly close when escape key focuses trigger because we prevented this
+        // in on focus outside for pointers.
+        context.onOpenChange(false);
+        context.trigger?.focus();
+      })}
+      onKeyDown={composeEventHandlers(props.onKeyDown, (event) => {
+        const element = event.target as HTMLElement;
+        // Submenu key events bubble through portals. We only care about keys in this menu.
+        const isKeyDownInside = event.currentTarget.contains(element);
+        const isCloseKey = SUB_CLOSE_KEYS[context.dir].includes(event.key);
+        if (isKeyDownInside && isCloseKey) {
+          // We need to explicitly close when keyboard focuses trigger because we prevented this
+          // in on focus outside for pointers.
```

Comment:
This is similar to the other one, I think I see why you wrote it like that now.
But I still find it odd to read.

Really, it's 2 things:
- we have to close manually because we're not wiring onto `onDismiss` in `onPointerDownOutside` and `onEscapeKeydown` (but crucially not in `onFocusOutside`)
- we need to focus manually because we have prevented it in `onCloseAutoFocus`

Reply:
I've simplified this one and the `onEscapeKeyDown` one to read just:

```
// We focus manually because we prevented it in `onCloseAutoFocus`.
```

We already explain why we prevent `onCloseAutoFocus` alongside that event so seems unnecessary to explain it again each time.